### PR TITLE
HDDS-4181. Add acceptance tests for upgrade, finalization and downgrade.

### PR DESCRIPTION
**Pull request by errose28**
## What changes were proposed in this pull request?

Robot acceptance test and shell scripting framework to automate upgrades and downgrades. Commands can be inserted at any stage in an upgrade/downgrade with callbacks. See README.md for more information. Downgrades will not be supported until the 1.1.0 to 1.2.0 upgrade process, so the downgrade tests have been commented out and will be added after the 1.1.0 release.

## What is the link to the Apache JIRA

HDDS-4181

## How was this patch tested?

Added acceptance tests should pass.

### CODE REVIEW

File: hadoop-ozone/dist/src/main/compose/upgrade/upgrades/non-rolling-upgrade/test.sh
Comment (avijayanhwx) [reply hadoop-ozone/dist/src/main/compose/upgrade/upgrades/non-rolling-upgrade/test.sh:92]: Cool, thanks!

**Review by avijayanhwx — APPROVED** [review]
Thanks for working on this @errose28. LGTM +1. Let's also wait for a review from @adoroszlai or @elek.

**Review by adoroszlai — COMMENTED** [review]
Thanks @errose28 for the patch.  Mostly looks good to me, but have some minor questions/comments.

File: hadoop-ozone/dist/src/main/compose/testlib.sh
Comment (errose28) [reply hadoop-ozone/dist/src/main/compose/testlib.sh:380]: Oh that makes sense. Definitely a copy error, I'll fix it.

File: hadoop-ozone/dist/src/main/compose/testlib.sh
Comment (errose28) [reply hadoop-ozone/dist/src/main/compose/testlib.sh:394]: Good catch, I'll take them out.

File: hadoop-ozone/dist/src/main/compose/upgrade/testlib.sh
Comment (errose28) [reply hadoop-ozone/dist/src/main/compose/upgrade/testlib.sh:72]: Sounds like a good idea. We are not expecting executable files or anything else here. I'll add it.

**Review by errose28 — COMMENTED** [review]
Thanks for the review and suggestions @adoroszlai. Let me know if there are still issues/questions with the new set of changes.

**Comment by errose28** [comment]
Acceptance test failure in most recent CI run was introduced after SCM HA merge and is fixed in HDDS-4828, which should be merged soon.

File: hadoop-ozone/dist/src/main/compose/upgrade/README.md
Comment (adoroszlai) [reply hadoop-ozone/dist/src/main/compose/upgrade/README.md:62]: Thanks for the explanation.  Then were good, the `with_new_version_finalized` callback covers the case I had in mind.

File: hadoop-ozone/dist/src/main/compose/upgrade/upgrades/manual-upgrade/test.sh
Comment (adoroszlai) [reply hadoop-ozone/dist/src/main/compose/upgrade/upgrades/manual-upgrade/test.sh:23]: > Maybe that function can be modified to read from a variable specifying the name of the file to use.

Or pass it as second, optional argument.

File: hadoop-ozone/dist/src/main/compose/upgrade/upgrades/manual-upgrade/test.sh
Comment (adoroszlai) [inline hadoop-ozone/dist/src/main/compose/upgrade/upgrades/manual-upgrade/test.sh:33]: Is the intention here to fail if the callback file does not exist?

Comment (errose28) [reply hadoop-ozone/dist/src/main/compose/upgrade/upgrades/manual-upgrade/test.sh:33]: No, actually the opposite. The callback file should be optional since all the functions in it are optional. This way the framework can be used to quickly try out an upgrade between two versions without defining an empty file if no callbacks are desired. The `callback` function used to invoke the callbacks will print a message for each undefined callback it skips.

File: hadoop-ozone/dist/src/main/compose/upgrade/upgrades/manual-upgrade/test.sh
Comment (errose28) [reply hadoop-ozone/dist/src/main/compose/upgrade/upgrades/manual-upgrade/test.sh:23]: Done

File: hadoop-ozone/dist/src/main/compose/upgrade/upgrades/manual-upgrade/test.sh
Comment (adoroszlai) [reply hadoop-ozone/dist/src/main/compose/upgrade/upgrades/manual-upgrade/test.sh]: But given `set -e`, I think the above would fail if the file does not exist.  To make it optional:

```bash
[[ ! -f "$OZONE_UPGRADE_CALLBACK" ]] || source "$OZONE_UPGRADE_CALLBACK"
```

(or use `if`).

Comment (errose28) [reply hadoop-ozone/dist/src/main/compose/upgrade/upgrades/manual-upgrade/test.sh]: I tested it without a callback file and it worked, but I have updated the sourcing lines to your suggestion to be safe.

Comment (adoroszlai) [reply hadoop-ozone/dist/src/main/compose/upgrade/upgrades/manual-upgrade/test.sh]: No, sorry, you are right:

> The shell does not exit if the command that fails is ... part of a && or || list

Comment (errose28) [reply hadoop-ozone/dist/src/main/compose/upgrade/upgrades/manual-upgrade/test.sh]: Ok thanks, I reverted the change since I think the original reads better and the functionality will be identical.

**Review by adoroszlai — APPROVED** [review]
Thanks a lot @errose28 for the detailed explanations.

**Comment by errose28** [comment]
Thank you @adoroszlai and @avijayanhwx for the reviews.

**Comment by avijayanhwx** [comment]
Thank you @errose28 for working on this, and @adoroszlai for the reviews. I am merging this.

**Pull request merged**
